xds/src/main/java/io/grpc/xds/ClusterResolverLoadBalancer.java

```diff
@@ -0,0 +1,684 @@
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    if (xdsClientPool == null) {
+      xdsClientPool = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_POOL);
+      xdsClient = xdsClientPool.getObject();
+    }
+    ClusterResolverConfig config =
+        (ClusterResolverConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    if (!Objects.equals(this.config, config)) {
+      logger.log(XdsLogLevel.DEBUG, "Config: {0}", config);
+      delegate.switchTo(new ClusterResolverLbStateFactory());
+    }
+    this.config = config;
+    delegate.handleResolvedAddresses(resolvedAddresses);
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    logger.log(XdsLogLevel.WARNING, "Received name resolution error: {0}", error);
+    delegate.handleNameResolutionError(error);
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    delegate.shutdown();
+    if (xdsClientPool != null) {
+      xdsClientPool.returnObject(xdsClient);
+    }
+  }
+
+  private final class ClusterResolverLbStateFactory extends LoadBalancer.Factory {
+    @Override
+    public LoadBalancer newLoadBalancer(Helper helper) {
+      return new ClusterResolverLbState(helper);
+    }
+  }
+
+  /**
+   * The state of a cluster_resolver LB working session. A new instance is created whenever
+   * the cluster_resolver LB receives a new config. The old instance is replaced when the
+   * new one is ready to handle new RPCs.
+   */
+  private final class ClusterResolverLbState extends LoadBalancer {
+    private final Helper helper;
+    private final List<String> clusters = new ArrayList<>();
+    private final Map<String, ClusterState> clusterStates = new HashMap<>();
+    // An aggregate cluster is thought of as a cluster that groups the endpoints of the underlying
+    // clusters together for load balancing purposes only. Load balancing policies (both locality
+    // level and endpoint level) are configured by the aggregate cluster and apply to all of its
+    // underlying clusters.
+    private PolicySelection localityPickingPolicy;
+    private PolicySelection endpointPickingPolicy;
+    private ResolvedAddresses resolvedAddresses;
+    private LoadBalancer childLb;
+
+    ClusterResolverLbState(Helper helper) {
+      this.helper = checkNotNull(helper, "helper");
+      logger.log(XdsLogLevel.DEBUG, "New ClusterResolverLbState");
+    }
+
+    @Override
+    public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+      this.resolvedAddresses = resolvedAddresses;
+      ClusterResolverConfig config =
+          (ClusterResolverConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+      for (DiscoveryMechanism instance : config.discoveryMechanisms) {
+        clusters.add(instance.cluster);
```

Comment:
Seems you assume the caller has de-duplicated the `resolvedAddresses`, not only `resolvedAddresses.getLoadBalancingPolicyConfig()`, so this code path can only be called once. This assumption on resolver and upstream lbs is too strong.

Reply:
What do you mean for `resolvedAddresses` is de-duplicated? It is assuming clusters in the DiscoveryMechanism list are unique. Isn't that a valid assumption?

This is in the switchable component of ClusterResolverLoadBalancer, which is instantiated for any `ClusterResolverConfig` changes. It's a very coarse-grained handling for state transition, and initially I did not like it as mentioned in the design as this causes the whole LB subtree to be changed every time a single cluster's configuration changes. But after trying out other approaches such as to handle each cluster's configuration changes individually inside the LB policy, there were some very subtle issues such as keeping each update atomic and consistent.

Reply:
>What do you mean for resolvedAddresses is de-duplicated?

By de-duplicated, I mean `handleResolvedAddresses()` only gets called once for `ClusterResolverLbState`. Yes, `resolvedAddresses.getLoadBalancingPolicyConfig()` is never changed for `ClusterResolverLbState`, but could `handleResolvedAddresses()` be called more than once? Why not call `delegate.handleResolvedAddresses(resolvedAddresses)` inside `if (!Objects.equals(this.config, config)) {}`

Reply:
I see, you mean superfluous updates. Fixed now.